File: bombies/OS_algorithms/sstf.cpp
// C++ program for implementation of
// SSTF disk scheduling
#include <algorithm>
#include <cmath>
#include <iostream>
#include <vector>
using namespace std;

void calculatedifference(int request[], int head, int diff[][2], int size);
int findMIN(int diff[][2], int size);
void shortestSeekTimeFirst(int request[], int head, int size);

//****************************************************************************************************

int main() {
    int size,
        head;
    // request array

    cout << "SSTF" << endl;
    cout << "How many request: ";
    cin >> size;
    cout << "Read/Write head starts on track: ";
    cin >> head;

    int* proc = new int[size];
    for (int i = 0; i < size; ++i) {
        cout << "Enter request " << i + 1 << ": ";
        cin >> proc[i];
    }

    shortestSeekTimeFirst(proc, head, size);
    delete[] proc;
    return 0;
}

//****************************************************************************************************

// Calculates difference of each
// track number with the head position
void calculatedifference(int request[], int head, int diff[][2], int size) {
    for (int i = 0; i < size; i++) {
        diff[i][0] = abs(head - request[i]);
    }
}

//****************************************************************************************************

// Find unaccessed track which is
// at minimum distance from head
int findMIN(int diff[][2], int size) {
    int index = -1;
    int minimum = 1e9;

    for (int i = 0; i < size; i++) {
        if (!diff[i][1] && minimum > diff[i][0]) {
            minimum = diff[i][0];
            index = i;
        }
    }
    return index;
}

//****************************************************************************************************

void shortestSeekTimeFirst(int request[], int head, int size) {
    if (size == 0) {
        return;
    }

    // Create array of objects of class node
    int diff[size][2] = {{0, 0}};

    // Count total number of seek operation
    int seekcount = 0;

    // Stores sequence in which disk access is done
    int seeksequence[size + 1] = {0};

    for (int i = 0; i < size; i++) {
        seeksequence[i] = head;
        calculatedifference(request, head, diff, size);
        int index = findMIN(diff, size);
        diff[index][1] = 1;

        // Increase the total count
        seekcount += diff[index][0];

        // Accessed track is now new head
        head = request[index];
    }
    seeksequence[size] = head;

    cout << "\nTotal number of seek operations = " << seekcount << endl;
    cout << "Average seek: " << (static_cast<double>(seekcount) / size) << endl;
    cout << "\nSeek sequence" << endl;

    // Print the sequence
    for (int i = 0; i <= size; i++) {
        cout << seeksequence[i] << "\n";
    }
}

/*

SSTF
How many request: 9
Read/Write head starts on track: 143
Enter request 1: 86
Enter request 2: 147
Enter request 3: 91
Enter request 4: 177
Enter request 5: 94
Enter request 6: 150
Enter request 7: 102
Enter request 8: 175
Enter request 9: 130

Total number of seek operations = 162
Average seek: 18

Seek sequence
143            // Read/Write head starts on track
147
150
130
102
94
91
86
175
177

*/

File: python/OS_algorithms/sstf.py
# Python3 program for implementation of
# SSTF disk scheduling

# Calculates difference of each
# track number with the head position
def calculateDifference(queue, head, diff):
    for i in range(len(diff)):
        diff[i][0] = abs(queue[i] - head)


# find unaccessed track which is
# at minimum distance from head
def findMin(diff):

    index = -1
    minimum = 999999999

    for i in range(len(diff)):
        if not diff[i][1] and minimum > diff[i][0]:
            minimum = diff[i][0]
            index = i
    return index


def shortestSeekTimeFirst(request, head):
    if len(request) == 0:
        return

    l = len(request)
    diff = [0] * l

    # initialize array
    for i in range(l):
        diff[i] = [0, 0]

    # count total number of seek operation
    seek_count = 0

    # stores sequence in which disk
    # access is done
    seek_sequence = [0] * (l + 1)

    for i in range(l):
        seek_sequence[i] = head
        calculateDifference(request, head, diff)
        index = findMin(diff)

        diff[index][1] = True

        # increase the total count
        seek_count += diff[index][0]

        # accessed track is now new head
        head = request[index]

    # for last accessed track
    seek_sequence[len(seek_sequence) - 1] = head

    print("Total number of seek operations =", seek_count)

    print("Seek Sequence is")

    # print the sequence
    for i in range(l + 1):
        print(seek_sequence[i])


# Driver code
if __name__ == "__main__":

    # request array
    proc = [86, 147, 91, 177, 94, 150, 102, 175, 130]
    head = 143
    shortestSeekTimeFirst(proc, head )
